[Tan Yan-Hao Joshua] iP

joshua-tyh opened the pull request:
# Sunday :wave:

>“Checkout my better half, Saturday on the [iOS App Store](https://apps.apple.com/sg/app/saturday-split-smarter/id1635430816).” – Team Saturday Devs

Sunday is less accomplished than Saturday on the [iOS App Store](https://apps.apple.com/sg/app/saturday-split-smarter/id1635430816), in that it lacks image-to-text capabilities. Instead it's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/nus-cs2103-AY2223S2/ip/pull/305).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you

And it is **FREE**, just like Saturday on the [iOS App Store](https://apps.apple.com/sg/app/saturday-split-smarter/id1635430816)! :star_struck:

Features:
- [x] Managing tasks (todos, deadlines & events)
- [x] Functional GUI
- [ ] Reminders (coming soon)
- [ ] Sleek GUI (coming not so soon)

Commit:
Add Gradle support

joshua-tyh committed:
Level-1

joshua-tyh committed:
Level-2

joshua-tyh committed:
Level-3

joshua-tyh committed:
A-Classes

joshua-tyh committed:
Level-4, A-Inheritance

joshua-tyh committed:
A-TextUiTesting

joshua-tyh committed:
A-Exceptions

joshua-tyh committed:
Update exception handling for MARK and UNMARK states

Reorganised and fixed the try-catch block for MARK and UNMARK states.

joshua-tyh committed:
Add DELETE state

Implemented a DELETE state for delete commands to remove tasks from the list. The NumberFormatException and IndexOutOfBoundsException are handled also accordingly.

joshua-tyh committed:
Remove emoticon in error messages

Error messages no longer have an emoticon to remedy the unmappable character (0xB9) for encoding US-ASCII error.

joshua-tyh committed:
Add DataHandler class to create, read and write data to save file.

joshua-tyh committed:
Teach Sunday how to understand dates and times.

Input format of dates and times follows "dd/mm/yyyy hhmm". Write format of dates and times follows "E, MMM dd yyyy, h:mm aa".

The text-ui-test inputs and expected outputs have also been updated accordingly.

joshua-tyh committed:
Resolve conflicts for merge banch-level-8.

Resolve using "mine" for Sunday.java, State.java and Printer.java files.

joshua-tyh committed:
Fix save bug for deadline and event tasks.

Deadline and Event tasks are now save with their date attributes according to the read format of ""dd/mm/yyyy hhmm". To achieve this, the Task class has been made abstract with the save() abstract method for child classes to override with their own implementations of how it should be saved.

A minor change was also made to the DataHandler class to remove the wildcard import. Redundant imports were also removed from the State class and Sunday class.

joshua-tyh committed:
Refactor code to extract out closely related code as classes.

Created the Parser and Ui classes. Renamed State.java to Command.java, Records.java to TaskList.java and DataHandler.java to Storage.java.

joshua-tyh committed:
Organize classes into suitable java packages

Packages created:
- collections
	- TaskList.java
- command
	- Command.java
- exceptions
	- SundayException.java
- task
	- Deadline.java
	- Event.java
	- Task.java
	- Todo.java
-utilities
	- Parser.java
	- Storage.java
	- Ui.java

joshua-tyh committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

joshua-tyh committed:
Set up gradle to automate some of the build tasks of the project

Set up gradle to build and run Sunday. Gradle has also been set up to run checkstyleMain with the checkstyle.xml file.

RubyNguyen07 commented on src/main/java/Sunday.java:
```diff
@@ -0,0 +1,33 @@
+                Ui.printException(e);
+            }
+        }
+        this.ui.close();
```
Do you think having another method for destroying the ui will make it easier to understand? 🤔

joshua-tyh replied:
This method was created to close the Scanner object that took inputs from the user. Apologies for the confusion!

RubyNguyen07 commented on src/main/java/Sunday.java:
```diff
@@ -0,0 +1,33 @@
+            }
+        }
+        this.ui.close();
+    }
+    public static void main(String[] args) {
```
Line break i think could be inserted here to improve readability

joshua-tyh replied:
You're right! I've added line breaks everywhere since this commit to improve readability. Thanks!

RubyNguyen07 commented on src/main/java/collections/TaskList.java:
```diff
@@ -0,0 +1,69 @@
+        int count = 0;
+        for (int i = 0; i < this.list.size(); i++) {
+            if (!this.list.get(i).isComplete()) {
+                count++;
+            }
+        }
+        return count;
+    }
+    public boolean save() throws SundayException {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < list.size(); i++) {
+            if (i > 0) {
+                sb.append("\n");
+            }
+            sb.append(this.list.get(i).save());
+        }
+        return Storage.writeToDataFile(sb.toString());
+    }
+    public boolean load() throws SundayException{
+        if (Storage.createDataFile()) {
+            return true;
+        }
+        Storage.readFromDataFile();
+        return false;
+    }
+    public boolean isEmpty() {
+        return this.list.isEmpty();
+    }
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < list.size(); i++) {
+            if (i > 0) {
+                sb.append("\n");
+            }
+            sb.append(this.getTask(i));
+        }
+        return sb.toString();
+    }
+}
```
Do you think placing line breaks between methods will help the structure look more logically?

RubyNguyen07 commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,184 @@
+            try {
+                input = input.substring(1);
+                String[] strArr = input.split(" ");
+                int j = 0;
+                StringBuilder sb = new StringBuilder();
+                while (!(strArr[j].equals("/from") || strArr[j].equals("(from:"))) {
+                    sb.append(strArr[j]);
+                    sb.append(" ");
+                    j++;
+                }
+                String description = sb.toString().substring(0, sb.length() - 1);
+
+                sb.setLength(0);
+                j++; // skip "/from" or "(from:"
+                while (!(strArr[j].equals("/to") || strArr[j].equals("to:"))) {
+                    sb.append(strArr[j]);
+                    sb.append(" ");
+                    j++;
+                }
+                String start = sb.toString().substring(0, sb.length() - 1);
+
+                sb.setLength(0);
+                j++; // skip "/to" or "to:"
+                while (j < strArr.length) {
+                    sb.append(strArr[j]);
+                    sb.append(" ");
+                    j++;
+                }
+                String end = sb.toString().substring(0, sb.length() - 1);
+                Task event = new Event(description, start, end);
+                list.add(event);
+                Ui.printAddedTask(event, list.getUncompletedSize());
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new SundayException("OOPS!!! The description of an event cannot be empty.");
+            } catch (ArrayIndexOutOfBoundsException e) {
+                throw new SundayException("OOPS!!! The description of this event is invalid.");
+            } catch (ParseException e) {
+                throw new SundayException(
+                        "OOPS! It appears the deadline given was not of the format dd/mm/yyyy hhmm");
+            }
```
As I notice there are a few lines of code that are similar to each other (have similar functionality), do you think creating another helper method to deal with duplicate lines will help reducing the number of code lines and improve  understandability?

joshua-tyh replied:
As this implementation is specific to the Event task, I didn't see a need to abstract away the similar lines of code, however I do agree that the duplicity of my code could affect understandability!

RubyNguyen07 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public String save() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[D]");
+        sb.append(super.toString()).append(" (by: ");
+        sb.append(super.readFormat.format(this.deadline));
+        sb.append(")");
```
This also seems similar to line 22-27 as well. Perhaps creating a helper method will help

joshua-tyh replied:
This was to ensure that the task was saved according to the read format and not the write format, hence the similar but importantly different method.

RubyNguyen07 reviewed (commented):
LGTM. Just a few nits to fix!

joshua-tyh committed:
Add JUnit Tests for all Task types.

JUnit Tests for all Task types test the creation and saving of each task type.

Detected a bug when saving Deadline and Event tasks where the saved date and time was incorrect. The bug was fixed by changing the DataFomat from "dd/mm/yyyy hhmm" to "dd/MM/yyyy HHmm".

joshua-tyh committed:
Fix relative path for save file

Fixed bug where save file was unable to be retrieved due to inaccurate relative file path. Text-Ui-Test files have also been updated.

joshua-tyh committed:
Fix relative path for save file

Added a getFilepath method in Storage to accurately retrieve the save file directory.

joshua-tyh committed:
Add JavaDoc comments and tweak the code to comply with a coding standard

joshua-tyh committed:
Revert "Add JavaDoc comments and tweak the code to comply with a coding standard"

This reverts commit 528c21a1904ddc3680bfc02b6d11e625ef6bc204.

joshua-tyh committed:
Add JavaDoc comments in A-JavaDoc branch

joshua-tyh committed:
Tweak the code to comply with a coding standard in A-CodingStandard branch

joshua-tyh committed:
Implement Find in Level-9 branch

Give users a way to find a task by searching for a keyword.

joshua-tyh committed:
Merge branch 'branch-A-JavaDoc'

joshua-tyh committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Sunday.java
#	src/main/java/collections/TaskList.java
#	src/main/java/command/Command.java
#	src/main/java/exceptions/SundayException.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java
#	src/main/java/task/Task.java
#	src/main/java/task/ToDo.java
#	src/main/java/utilities/Parser.java
#	src/main/java/utilities/Storage.java
#	src/main/java/utilities/Ui.java

joshua-tyh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/collections/TaskList.java
#	src/main/java/command/Command.java
#	src/main/java/task/Task.java

joshua-tyh committed:
Amend checkstyle violations

Used Gradle to check for checkstyle violations and amended them accordingly.

joshua-tyh committed:
Implement GUI to Sunday using JavaFX.

Update Sunday behaviour to take user inputs from GUI instead of the console.

ChangGittyHub commented on src/main/java/utilities/Ui.java:
```diff
@@ -123,4 +123,17 @@ public static void printEmptyTaskList() {
         printText("Tell me what to note down and I'll remember it accordingly!");
         printBar();
     }
+    public static void printListFound(TaskList taskList) {
```
I liked that you tried to make the name descriptive but printMatchedList might be clearer?

joshua-tyh replied:
This method has since been deprecated but thanks for the input!

ChangGittyHub commented on src/main/java/task/Task.java:
```diff
@@ -24,6 +24,9 @@ public void unmark() {
+    }
     @Override
     public String toString() {
         StringBuilder sb = new StringBuilder();
```
do you think it might be better to have a descriptive name in english instead of short form?

joshua-tyh replied:
Right! I've updated the method to make it easier to read. Thanks for the input!

ChangGittyHub commented on src/main/java/collections/TaskList.java:
```diff
@@ -55,6 +55,16 @@ public boolean load() throws SundayException{
+        for (int i = 0; i < this.list.size(); i++) {
+            Task task = this.list.get(i);
+            if (task.match(keyword)) {
+                found.add(task);
```
Do you think this might be bit confusing? Maybe Tasklist instance can be called matchedList or foundList? the result could look like this: matchedList.add(task) which might be clearer

ChangGittyHub commented on src/main/java/Command.java:
```diff
@@ -169,24 +154,15 @@ public void execute(String input) throws SundayException{
-            Printer.printBar();
+        public void execute(String input) throws SundayException {
+            boolean didSave = list.save();
+            Ui.printGoodbye(didSave);
```
maybe isSaved might be better?

ChangGittyHub commented on src/main/java/DataHandler.java:
```diff
@@ -0,0 +1,81 @@
+                //strArr[1] = "[" or "[X"
+                //strArr[2] = " \(description)"
+
+                switch (strArr[0].charAt(1)) {
```
could this be abstracted out?

ChangGittyHub commented on src/main/java/DataHandler.java:
```diff
@@ -0,0 +1,81 @@
+                //strArr[2] = " \(description)"
+
+                switch (strArr[0].charAt(1)) {
+                case 'T':
```
I like how you remember the right indentation for switch case

ChangGittyHub commented on src/main/java/State.java:
```diff
@@ -50,37 +66,39 @@ public void execute(String input) throws SundayException{
                 sb.setLength(0);
-                while (!strArr[j].equals("/to")) {
+                j++; // skip "/from" or "(from:"
+                while (!(strArr[j].equals("/to") || strArr[j].equals("to:"))) {
```
strArr[j] could be abstracted out for better readability?

ChangGittyHub commented on src/main/java/Printer.java:
```diff
@@ -1,10 +1,15 @@
 public class Printer {
-    private final String bar = "____________________________________________________________";
```
could this be better if it was a final String named BAR instead of the change you made?

ChangGittyHub reviewed (commented):
LGTM.Just some nits to fix

joshua-tyh committed:
Amend files for CheckStyle violations

Used Gradle to detect checkstyle violations and fixed them accordingly. Changes are made in a A-CheckStyle branch.

joshua-tyh committed:
Add a GUI to Sunday

Use the JavaFX technology to implement the GUI. Commit changes to the Level-10 branch

joshua-tyh committed:
Revert "Implement GUI to Sunday using JavaFX."

This reverts commit 5aaedfabc8f37e2fe0ab9a08e75b015c79288c63.

joshua-tyh committed:
Merge branch 'branch-A-CheckStyle'

joshua-tyh committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/utilities/Ui.java

joshua-tyh committed:
Fix Command enums to return messages instead of printing to the console

joshua-tyh committed:
Change build.gradle for new JAR file

Fixed the build.gradle file for the task shadowJar to create a JAR file.

joshua-tyh committed:
Add assert statements at various points in the code

Assert statements were added to document important assumptions that
should hold at various points in the code.

Let's add assert statements in the following,
* In the DEADLINE command to check that the deadline description
contains a "/by" or "(by:" field.
* In the EVENT command to check that the event description contains a
"/from" or "(from:" field.
* In the EVENT command to check that the event description contains a
"/to" or "to:" field.
* In the TODO command to check that the todo description is not empty.
* In the add(Task task) method to check that the task parameter is not
null.
* In the mark(int index) method to check that the index parameter is
within the bounds of the list.
* In the unmark(int index) method to check that the index parameter is
within the bounds of the list.
* In the delete(int index) method to check that the index parameter is
within the bounds of the list.
In the getTask	(int index) method to check that the index parameter is
within the bounds of the list.

joshua-tyh committed:
Improve code quality

A few methods throughout the codebase were not being called/utilised.

Removing all those redundant methods improves the code quality.

Let's remove those redundant methods and cleanup the codebase to improve
the code quality.

joshua-tyh committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/command/Command.java

joshua-tyh committed:
Add a Sort feature

To give proper ordering to the list of tasks, a sort method should be
added to allow the user to sort the tasks in the list.

Sorting the tasks in alphabetical order according to the task description
will enable the user to better locate tasks in the list.

To implement this feature, let's:
* Implement the Comparable interface in the Task class and override the
compareTo() method to compare task descriptions.
* Add a "Sort" command for the user to invoke the sort feature.

joshua-tyh committed:
Comply with checkstyle requirements

Updated codebase to comply with checkstyle requirements according to
gradle checkstyleMain.

joshua-tyh committed:
Add a User Guide

Add a representative screenshot of the product to the docs folder named
"Ui.png".

joshua-tyh committed:
Remove emoji from README.md

Emojis were not showing up on the GitHub Pages view. Removing them to
improve readability.

joshua-tyh committed:
Update README.md

joshua-tyh committed:
Add javadoc for getResponse() in Sunday.java

joshua-tyh committed:
Update javadoc in Sunday.java

joshua-tyh committed:
Merge pull request #2 from joshua-tyh/branch-A-Assertions

Add javadoc for getResponse() in Sunday.java

joshua-tyh committed:
Merge branch 'master' into branch-A-CodeQuality

joshua-tyh committed:
Merge pull request #3 from joshua-tyh/branch-A-CodeQuality

Branch a code quality
